Feature: Reverse Vouchers

jniles opened the pull request:
This PR implements the reverse voucher feature.  It is functionally identical to the other reversals, using the `CREDIT_NOTE` transaction type (which should probably now be called correction).  When a row is reversed, it turns red like the other registries and a call to `ReverseTransaction()` is made.

Additionally, a bug is fixed with cash reversals that show up with empty descriptions.  This was caused by an incorrect binding.

Closes #1204.  Closes #2927.

**Screenshots**
![voucherreverserecorddropdown](https://user-images.githubusercontent.com/896472/41870324-a6db2632-78b3-11e8-8bcd-872542abbb7a.PNG)
_Fig 1: Reverse Record in the Dropdown_

![voucherreverserecordmodal](https://user-images.githubusercontent.com/896472/41870325-a7241a0e-78b3-11e8-96bc-9e6c95aa6f79.PNG)
_Fig 2: Reverse Record Modal_

![voucherreverserecordrow](https://user-images.githubusercontent.com/896472/41870326-a75f77de-78b3-11e8-97c6-c8f074b929a0.PNG)
_Fig 3: A reversed row in the Voucher Registry_

jniles committed:
feat(vouchers): implement reverse record

This commit implements the client-side for voucher reversals.

jniles committed:
feat(vouchers): add reverse voucher

This commit adds the server side for reversing vouchers. Functionally,
it is identical to reversing the cash payments or invoices, using the
same API.  It also consolidates the grid row templates for all
registries that have reversals into a single template to make things
easier to maintain.

Finally, a bug is fixed with cash reversals that show up with empty
descriptions.

Closes #1204.  Closes #2927.

mbayopanda commented on client/src/modules/vouchers/vouchers.service.js:
```diff
@@ -142,14 +143,16 @@ function VoucherService(
+   * This method reverses a transaction.
+   * bhima should automatically be able to reverse any transaction in the
+   * posting_journal by creating a new transaction that is an exact duplicate of
+   * the original transaction with the debits and credits switched.
```
👍

mbayopanda commented on server/controllers/finance/vouchers.js:
```diff
@@ -370,6 +371,10 @@ function safelyDeleteVoucher(guid) {
 
+  const TOGGLE_VOUCHER_REVERSAL = `
+      UPDATE voucher SET voucher.reversed = 0 WHERE voucher.uuid = ?;
+  `;
```
Is is correct to have `SET voucher.reversed = 0` for the `TOGGLE_VOUCHER_REVERSAL` instead of `SET voucher.reversed = 1`

jniles replied:
Yeah, I can see how this is confusing.

The code here is trying to flip the `reversed` flag if the reversing record (i.e. the voucher that made the reverse) is deleted.  So, if you delete the credit note, it will automatically remove the `reversed=1` from the invoice.   The same thing works here for vouchers.

mbayopanda commented on client/src/i18n/en/form.json:
```diff
@@ -60,6 +60,7 @@
             "RESET": "Reset",
             "RESET_DEFAULTS": "Reset Defaults",
             "RETURN": "Return",
+            "REVERSE_RECORD" : "Reverse Record",
```
The french translation is missing

jniles replied:
Could you give me one?

mbayopanda replied:
We can use `Inverser l'enregistrement`

jniles committed:
fix(invoices): consider prepayments before reverse

This commit changes the logic of the credit note field to check for
prepayments before reversing an invoice.  It will similarly display a
message about how the invoice has already been paid.

jniles committed:
fix(invoices): block credit notes if paid

This commit forces a user to reverse all associated payments before they
are allowed to reverse an invoice.

Closes #2931.

mbayopanda commented on server/controllers/finance/cash.js:
```diff
@@ -253,19 +255,28 @@ function update(req, res, next) {
+
+  const getInvoicePayment = `
+    SELECT DISTINCT BUID(cash_item.cash_uuid) cash_uuid, BUID(cash_item.invoice_uuid) invoice_uuid, cash.reversed
+    FROM cash JOIN cash_item ON cash.uuid = cash_item.cash_uuid
```
👍 good catch

mbayopanda reviewed (approved):
LGTM :+1

And it works fine

bors r+
